ext/ScreenShot/ScreenShot.h

```diff
@@ -5,27 +5,151 @@
+					imgViewInfo.viewType = asset::IImageView<asset::ICPUImage>::ET_2D;
+					imgViewInfo.flags = static_cast<asset::ICPUImageView::E_CREATE_FLAGS>(0u);
+					imgViewInfo.subresourceRange.baseArrayLayer = 0u;
+					imgViewInfo.subresourceRange.baseMipLevel = 0u;
+					imgViewInfo.subresourceRange.layerCount = imgInfo.arrayLayers;
+					imgViewInfo.subresourceRange.levelCount = imgInfo.mipLevels;
+
+					auto imageView = asset::ICPUImageView::create(std::move(imgViewInfo));
+					auto gpuImageView = driver->getGPUObjectsFromAssets(&imageView.get(), &imageView.get() + 1)->front();
+
+					return std::move(gpuImageView);
+				};
+
+				auto gpuImageViewDepthBuffer = createAttachement(false);
+				auto gpuImageViewColorBuffer = createAttachement(true);
+
+				auto frameBuffer = driver->addFrameBuffer();
+				frameBuffer->attach(video::EFAP_DEPTH_ATTACHMENT, std::move(gpuImageViewDepthBuffer));
+				frameBuffer->attach(video::EFAP_COLOR_ATTACHMENT0, std::move(gpuImageViewColorBuffer));
+
+				return frameBuffer;
+			};
+
+			/*
+				Create a ScreenShot with gpu image usage.
+			*/
+
+			bool createScreenShoot(core::smart_refctd_ptr<IrrlichtDevice> device, core::smart_refctd_ptr<video::IGPUImage> gpuImage, const std::string& outFileName)
+			{
+				auto driver = device->getVideoDriver();
+				auto assetManager = device->getAssetManager();
+
+				auto fetchedImageParams = gpuImage->getCreationParameters();
+				auto image = asset::ICPUImage::create(std::move(fetchedImageParams));
+
+				auto trueBufferRowLength = asset::IImageAssetHandlerBase::calcPitchInBlocks(fetchedImageParams.extent.width, asset::getTexelOrBlockBytesize(fetchedImageParams.format));
+
+				auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1u);
+				asset::ICPUImage::SBufferCopy& region = regions->front();
+
+				region.imageSubresource.mipLevel = 0u;
+				region.imageSubresource.baseArrayLayer = 0u;
+				region.imageSubresource.layerCount = 1u;
+				region.bufferOffset = 0u;
+				region.bufferRowLength = trueBufferRowLength;
+				region.bufferImageHeight = 0u;
+				region.imageOffset = { 0u, 0u, 0u };
+				region.imageExtent = image->getCreationParameters().extent;
+
+				auto destinationBuffer = core::smart_refctd_ptr<video::IGPUBuffer>(driver->createDownStreamingGPUBufferOnDedMem(image->getImageDataSizeInBytes()));
+				destinationBuffer->getBoundMemory()->mapMemoryRange(video::IDriverMemoryAllocation::EMCAF_READ, { 0u, destinationBuffer->getSize() });
+
+				driver->copyImageToBuffer(gpuImage.get(), destinationBuffer.get(), 1, regions->begin());
+
+				auto texelBuffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(image->getImageDataSizeInBytes());
+				auto rawData = reinterpret_cast<uint8_t*>(destinationBuffer->getBoundMemory()->getMappedPointer());
+
+				memcpy(texelBuffer->getPointer(), rawData, image->getImageDataSizeInBytes());
+
+				image->setBufferAndRegions(std::move(texelBuffer), regions);
```

Comment:
actually you can do a much more nasty thing here (avoid the memcpy).

Get the default down-stream from the driver, allocate an address in it, `copyImageToBuffer` to the offset given by the address, create an `asset::CCustomAllocatorCPUBuffer` over the memory address given by the mapped pointer with a null allocator and `adopt_memory` in the constructor (this way it won't allocate new memory, memcpy the contents, or free the pointer to the memory after it dies cause the null allocator), use that custom allocator CPUBuffer as backing for an `ICPUImage`, wait on the fence, pass the ICPUImage to the asset writer.

P.S. See the command-line Denoiser example, it uses that method to download data from OptiX denoiser, and make an ICPUImage out of it to save to OpenEXR

Reply:
not sure if it's done correctly now, but changed it